src/runtime-rs/crates/hypervisor/src/qemu/network.rs

```diff
@@ -0,0 +1,272 @@
+    type Error = anyhow::Error;
+
+    fn try_from(device_driver: &str) -> Result<Self> {
+        Ok(match device_driver {
+            "virtio-net-pci" => DeviceDriver::VirtioNetPci,
+            "vfio-pci" => DeviceDriver::VfioPci,
+            _ => return Err(anyhow!("unsupported transport")),
+        })
+    }
+}
+
+#[derive(Debug, Default, PartialEq)]
+enum NetDev {
+    /// Tap("tap") is a tap networking device type.
+    #[default]
+    Tap,
+
+    /// MacTap("macvtap") is a macvtap networking device type.
+    #[allow(dead_code)]
+    MacvTap,
+}
+
+impl ToString for NetDev {
+    fn to_string(&self) -> String {
+        match self {
+            NetDev::Tap | NetDev::MacvTap => "tap".to_owned(),
+            // VhostUser is to be added in future.
+            // NetDev::VhostUser => "vhost-user".to_owned(),
+        }
+    }
+}
+
+// NetDevice represents a guest networking device
+// -netdev tap,id=hostnet0,vhost=on,vhostfds=x:y:z,fds=a:b:c
+// -device virtio-net-pci,netdev=hostnet0,id=net0,mac=24:42:54:20:50:46,bus=pci.0,addr=0x7
+#[derive(Debug, Default)]
+pub struct NetDevice {
+    // device_type is the netdev type (e.g. tap).
+    device_type: NetDev,
+
+    // driver is the qemu device driver
+    device_driver: DeviceDriver,
+
+    // id is the net device identifier.
+    id: String,
+
+    // if_name is the interface name,
+    if_name: String,
+
+    // bus is the bus path name of a PCI device.
+    bus: String,
+
+    // pci_addr is the address offset of a PCI device.
+    pci_addr: String,
+
+    // fds represents the list of already existing file descriptors to be used.
+    // This is mostly useful for mq support.
+    // {
+    //      fds:      Vec<File>,
+    //      vhost_fds: Vec<File>,
+    // }
+    fds: HashMap<String, Vec<RawFd>>,
+
+    // disable_vhost_net disables virtio device emulation from the host kernel instead of from qemu.
+    disable_vhost_net: bool,
+
+    // mac_address is the networking device interface MAC address.
+    mac_address: Address,
+
+    // disable_modern prevents qemu from relying on fast MMIO.
+    disable_modern: bool,
+
+    // transport is the virtio transport for this device.
+    transport: VirtioTransport,
+}
+
+impl NetDevice {
+    #[allow(dead_code)]
+    pub fn new(
+        config: &NetworkConfig,
+        disable_vhost_net: bool,
+        tun_fds: Vec<i32>,
+        vhost_fds: Vec<i32>,
+    ) -> Self {
+        // we have only two <Key, Value>s:
+        // {
+        //      "fds": vec![fd1, fd2,...],
+        //      "vhostfds": vec![fd3, fd4,...],
+        // }
+        let mut fds: HashMap<String, Vec<RawFd>> = HashMap::with_capacity(2);
+        fds.insert("fds".to_owned(), tun_fds);
+        fds.insert("vhostfds".to_owned(), vhost_fds);
+
+        // FIXME(Hard Code): It's safe to unwrap here because of the valid input.
+        let device_driver = DeviceDriver::try_from("virtio-net-pci").unwrap();
+        let transport = VirtioTransport::try_from("pci").unwrap();
```

Comment:
Ideally `device_driver` should be derived from `transport` to avoid code duplication.

At some point in time we might want to support other transports, e.g. s390x's CCW, but I guess it is acceptable to get started with PCI only.